### ISSUE TITLE
VS 15.4.0 cannot execute remote compiled binary

### ISSUE THREAD

**Issue by fjeremic** [Microsoft/VSLinux#193]
I'm doing something a little unorthodox. I'm attempting to follow the instructions [from the official blog post](https://blogs.msdn.microsoft.com/vcblog/2017/08/25/visual-c-for-linux-development-with-cmake/) to compile the sample `hello.cpp` program on a Linux on z (s390x) machine. Everything more or less works and VS is able to copy the files to the remote machine and correctly compile them. The only thing that is not working is that I cannot run the compiled program from VS.

Here is the build output. I first clean the solution, clean the CMake cache, then rebuild all, and try to Start (F5) the remote process:

```
Cleaning /tmp/f389b1b3-307d-4732-802c-085a64b262f5/build/Linux-Debug ...
Cleaning done.

1> 5:25:49 PM: Copying files to remote machine...
1> 5:25:49 PM: Finished copying files (elapsed time 00h:00m:00s:000ms).
1> /usr/local/bin/cmake -G "Unix Makefiles"  -DCMAKE_BUILD_TYPE="Debug" "\tmp\f389b1b3-307d-4732-802c-085a64b262f5\src\Linux-Debug"
1> -- The C compiler identification is GNU 5.4.0
1> -- The CXX compiler identification is GNU 5.4.0
1> -- Check for working C compiler: /usr/bin/cc
1> -- Check for working C compiler: /usr/bin/cc -- works
1> -- Detecting C compiler ABI info
1> -- Detecting C compiler ABI info - done
1> -- Detecting C compile features
1> -- Detecting C compile features - done
1> -- Check for working CXX compiler: /usr/bin/c++
1> -- Check for working CXX compiler: /usr/bin/c++ -- works
1> -- Detecting CXX compiler ABI info
1> -- Detecting CXX compiler ABI info - done
1> -- Detecting CXX compile features
1> -- Detecting CXX compile features - done
1> -- Configuring done
1> -- Generating done
1> -- Build files have been written to: /tmp/f389b1b3-307d-4732-802c-085a64b262f5/build/Linux-Debug
1> Starting CMake target info extraction ...
1> Extracted source files and headers.
1> Extracted global settings.
1> Extracted code model.
1> Collating data ...
1> Target info extraction done.

5:27:03 PM: Copying files to remote machine...
5:27:03 PM: Finished copying files (elapsed time 00h:00m:00s:000ms).
Scanning dependencies of target hello-cmake
[ 50%] Building CXX object CMakeFiles/hello-cmake.dir/hello.cpp.o
[100%] Linking CXX executable hello-cmake
[100%] Built target hello-cmake
```

This is what I get:

![image](https://user-images.githubusercontent.com/16259104/31253364-9da0e290-a9f2-11e7-89e4-2ef51bc42a7b.png)

It seems as if VS was unable to determine that this program is actually on a remote machine judging from the "${buildRoot}" prefix on the startup item. Neither of the two options work:

![image](https://user-images.githubusercontent.com/16259104/31253432-df1e8b64-a9f2-11e7-8158-ddf58fb94c90.png)

My `CMakeLists.txt` and `hellp.cpp` is the same as on the blog post. I have modified my `CMakeSettings.json` as such:

```
    {
      "name": "Linux-Debug",
      "generator": "Unix Makefiles",
      "remoteMachineName": "${defaultRemoteMachineName}",
      "configurationType": "Debug",
      "remoteCMakeListsRoot": "/tmp/${workspaceHash}/src/${name}",
      "cmakeExecutable": "/usr/local/bin/cmake",
      "buildRoot": "${env.USERPROFILE}\\CMakeBuilds\\${workspaceHash}\\build\\${name}",
      "installRoot": "${env.USERPROFILE}\\CMakeBuilds\\${workspaceHash}\\install\\${name}",
      "remoteBuildRoot": "/tmp/${workspaceHash}/build/${name}",
      "remoteInstallRoot": "/tmp/${workspaceHash}/install/${name}",
      "remoteCopySources": true,
      "remoteCopySourcesOutputVerbosity": "Normal",
      "remoteCopySourcesConcurrentCopies": "10",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "linux-x64" ]
    },
```

I can see the on my remote machine the program was successfully compiled, and I can execute it manually:

```
/tmp/f389b1b3-307d-4732-802c-085a64b262f5/build/Linux-Debug
> ll
total 64
drwxr-xr-x 3 fjeremic fjeremic  4096 Oct  5 17:25 ..
-rw-r--r-- 1 fjeremic fjeremic  1432 Oct  5 17:25 cmake_install.cmake
-rw-r--r-- 1 fjeremic fjeremic 12862 Oct  5 17:25 CMakeCache.txt
-rw-rw-r-- 1 fjeremic fjeremic  4983 Oct  5 17:25 Makefile
drwxr-xr-x 3 fjeremic fjeremic  4096 Oct  5 17:27 .
-rwxr-xr-x 1 fjeremic fjeremic 20903 Oct  5 17:27 hello-cmake
drwxr-xr-x 5 fjeremic fjeremic  4096 Oct  5 17:27 CMakeFiles
(xxxxxxx@fjeremic) [!]:
/tmp/f389b1b3-307d-4732-802c-085a64b262f5/build/Linux-Debug
> ./hello-cmake
Hello
```

Visual Studio Professional 2017 Preview 4 (15.4.0) Version 4.7.02046 (downloaded the latest yesterday)
Linux xxxxxxx 2.6.32-573.18.1.el6.s390x #1 SMP Wed Jan 6 11:16:09 EST 2016 s390x s390x s390x GNU/Linux
g++ (GCC) 5.4.0
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-83.el6)
GNU gdbserver (GDB) Red Hat Enterprise Linux (7.2-83.el6)

Server mode is supported on my CMake 3.9.0:

```
> cmake -E capabilities
{"generators":[{"extraGenerators":[],"name":"Watcom WMake","platformSupport":false,"toolsetSupport":false},{"extraGenerators":["CodeBlocks","CodeLite","Sublime Text 2","Kate","Eclipse CDT4"],"name":"Ninja","platformSupport":false,"toolsetSupport":false},{"extraGenerators":["CodeBlocks","CodeLite","Sublime Text 2","Kate","Eclipse CDT4","KDevelop3"],"name":"Unix Makefiles","platformSupport":false,"toolsetSupport":false}],"serverMode":true,"version":{"isDirty":false,"major":3,"minor":9,"patch":0,"string":"3.9.0","suffix":""}}
```

Is there some additional tracing I can enable to diagnose why VS is unable to launch the remote process? I can provide any other information needed to help debug this.

**Comment by fjeremic** [comment]
Running from the CMake menu via "Run Tests" -> "CMakeLists.txt (project hello-cmake)" does not work either:

```

Build succeeded.
Starting all tests..."C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\PREVIEW\PROFESSIONAL\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\ctest.exe" -C "Debug"   
Tests canceled. Exception thrown The directory name is invalid at    at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at libcmake.CMakeLists.<RunTestsAsync>d__106.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.VisualStudio.CMake.Project.FileScannerCache.<RunTests>d__40.MoveNext().
```

**Comment by lukka** [comment]
@fjeremic Thanks for the report. This bug has been fixed in VS 15.5 preview 1, which has not been released yet. ETA not available, please keep an eye on [here](https://www.visualstudio.com/en-us/news/releasenotes/vs2017-Preview-relnotes) for news about 15.5.

**Comment by fjeremic** [comment]
Just confirming VS 15.5 preview 1 fixes this issue. I am able to remotely launch but it seems I am unable to set breakpoints yet. I'll do some self exploration first and open up another issue if warranted. Thanks for fixing this!
